tests/integration/test_dirbrowser.py

```diff
@@ -0,0 +1,184 @@
+
+import os
+import bs4
+import collections
+
+import pytest
+
+from PyQt5.QtCore import QUrl
+from qutebrowser.utils import jinja
+
+
+class DirLayout:
+
+    """Provide a fake directory layout to test dirbrowser."""
+
+    Parsed = collections.namedtuple('Parsed', 'path, parent, folders, files')
+    Item = collections.namedtuple('Item', 'path, link, text')
+
+    LAYOUT = [
+        'folder0/file00',
+        'folder0/file01',
+        'folder1/folder10/file100',
+        'folder1/file10',
+        'folder1/file11',
+        'file0',
+        'file1',
+    ]
+
+    @classmethod
+    def layout_folders(cls):
+        """Return all folders in the root directory of the layout."""
+        folders = set()
+        for path in cls.LAYOUT:
+            parts = path.split('/')
+            if len(parts) > 1:
+                folders.add(parts[0])
+        folders = list(folders)
+        folders.sort()
+        return folders
+
+    @classmethod
+    def get_folder_content(cls, name):
+        """Return (folders, files) for the given folder in the root dir."""
+        folders = set()
+        files = set()
+        for path in cls.LAYOUT:
+            if not path.startswith(name + '/'):
+                continue
+            parts = path.split('/')
+            if len(parts) == 2:
+                files.add(parts[1])
+            else:
+                folders.add(parts[1])
+        folders = list(folders)
+        folders.sort()
+        files = list(files)
+        files.sort()
+        return (folders, files)
+
+    def __init__(self, factory):
+        self._factory = factory
+        self.base = factory.getbasetemp()
+        self.layout = factory.mktemp('layout')
+        self._mklayout()
+
+    def _mklayout(self):
+        for filename in self.LAYOUT:
+            self.layout.ensure(filename)
+
+    def file_url(self):
+        """Return a file:// link to the directory."""
+        return jinja.file_url(str(self.layout))
+
+    def path(self, *parts):
+        """Return the path to the given file inside the layout folder."""
+        return os.path.normpath(str(self.layout.join(*parts)))
+
+    def base_path(self):
+        """Return the path of the base temporary folder."""
+        return os.path.normpath(str(self.base))
+
+    def parse(self, quteproc):
+        """Parse the dirbrowser content from the given quteproc.
+
+        Args:
+            quteproc: The quteproc fixture.
+        """
+        html = quteproc.get_content(plain=False)
+        soup = bs4.BeautifulSoup(html, 'html.parser')
+        print(soup.prettify())
+        title_prefix = 'Browse directory: '
+        # Strip off the title prefix to obtain the path of the folder that
+        # we're browsing
+        path = soup.title.string[len(title_prefix):]
+        path = os.path.normpath(path)
+
+        container = soup('div', id='dirbrowserContainer')[0]
+
+        parent_elem = container('ul', class_='parent')
+        if len(parent_elem) == 0:
```

Comment:
nitpick: I think this can simply be `if not parent_elem:` as it should evaluate to false? Though I'm not sure what to expect with beautifulsoup :laughing: